### PR TITLE
FABGW-20 SDK support for endorsing orgs

### PR DESCRIPTION
**Pull request by andrew-coleman**
Added ability to specify endorsing orgs to the 3 SDKs.
Added new scenario test and implementation in the 3 SDKs.

Signed-off-by: andrew-coleman <andrew_coleman@uk.ibm.com>

### CODE REVIEW

**Review by jt-nti — DISMISSED** [review]

File: node/src/transaction.test.ts [hunk 1]
```diff
@@ -94,6 +94,23 @@ describe('Transaction', () => {
...
+            actualOrgs = req.endorsing_organizations;
+            return Promise.resolve({
+                prepared_transaction: {
+                    payload: Buffer.from('PAYLOAD'),
+                },
+                result: {
+                    payload: Buffer.from(expectedResult),
+                },
+            });
+        })
```
Comment (bestbeforetoday) [inline node/src/transaction.test.ts:109]: `client.endorse` is already a correctly configured mock function so you don't need to redefine it. After the call you can grab the request argument the mock captures with
```
const actualOrgs = client.endorse.mock.calls[0][0].endorsing_organizations;
```

File: scenario/features/privatedata.feature [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        Given I have deployed a Fabric network
+        And I have created and joined all channels
+        And I deploy golang chaincode named basic at version 1.0.0 for all organizations on channel mychannel with endorsement policy OR("Org1MSP.member","Org2MSP.member","Org3MSP.member")
+        And I create a gateway for user User1 in MSP Org1MSP
+        And I connect the gateway to peer0.org1.example.com
+        And I use the mychannel network
+        And I use the basic contract
+
+    Scenario: Submit to org1
+        When I prepare to submit an getPeerOrg transaction
```
Comment (bestbeforetoday) [inline scenario/features/privatedata.feature:17]: I hate to be that grammar lawyer but... "submit **a** getPeerOrg transaction"? 😉

File: scenario/fixtures/chaincode/golang/basic/main.go [hunk 3]
```diff
@@ -77,3 +109,12 @@ func toStringValues(input map[string][]byte) map[string]string {
...
 }
+
+func contains(slice []string, entry string) bool {
+	for _, item := range slice {
+		if entry == item {
+			return true
+		}
+	}
+	return false
+}
```
Comment (bestbeforetoday) [inline scenario/fixtures/chaincode/golang/basic/main.go:120]: I don't see this used anywhere. Could it be omitted?

File: scenario/node/src/fabric.ts [hunk 4]
```diff
@@ -209,8 +228,8 @@ export class Fabric {
             '--tlsRootCertFiles',
             '/etc/hyperledger/configtx/crypto-config/peerOrganizations/org2.example.com/peers/peer0.org2.example.com/tls/ca.crt');
 
-        this.runningChaincodes[mangledName] = true;
-        await sleep(10000);
+        this.runningChaincodes[mangledName] = signaturePolicy;
+        await sleep(30000);
```
Comment (bestbeforetoday) [inline scenario/node/src/fabric.ts:232]: For another day, but we really should find something better than this. Maybe polling the command to list running chaincodes until it appears?

**Review by bestbeforetoday — CHANGES_REQUESTED** [review]
Generally looking good. A few code tweaks I think are worth addressing.

I'm not sure if it's a good idea to keep adding more things to the _basic_ chaincode. Maybe we should have a different chaincode for specific scenarios like private data? It might (or might not!) also reduce the work redeploying because of different endorsement policies. We can figure that out in another PR though.

File: java/src/main/java/org/hyperledger/fabric/client/Proposal.java [hunk 5]
```diff
@@ -73,6 +73,16 @@
          */
         Builder putTransient(String key, byte[] value);
 
+        /**
+         * Specifies the set of organizations that will attempt to endorse the proposal.
+         * No other organizations' peers will be sent this proposal.
```
[the remark this answers is not in the document]
Comment (andrew-coleman) [reply java/src/main/java/org/hyperledger/fabric/client/Proposal.java:78]: This describes how the gateway behaves at the moment.  It overrides the endorsement policy and ensures that private data won't leak to other orgs.  The requirements for state-based endorsement will be addressed in a future user story.

File: node/src/transaction.test.ts (hunk 1 above)
Comment (andrew-coleman) [reply node/src/transaction.test.ts:109]: Thanks - that was the magic I was looking for :)

File: scenario/features/privatedata.feature (hunk 2 above)
Comment (andrew-coleman) [reply scenario/features/privatedata.feature:17]: quite right :)

File: scenario/fixtures/chaincode/golang/basic/main.go (hunk 3 above)
Comment (andrew-coleman) [reply scenario/fixtures/chaincode/golang/basic/main.go:120]: Yes, it got left in there from an earlier experiment

File: scenario/node/src/fabric.ts (hunk 4 above)
Comment (andrew-coleman) [reply scenario/node/src/fabric.ts:232]: I wonder if these sleeps are carried over from the old chaincode lifecycle.  We are now specifying `--waitForEvent` on the commit so hopefully these sleeps aren't needed at all now.  I've taken them out and it works locally, so see what it does in the CI.

Comment (andrew-coleman) [reply scenario/node/src/fabric.ts]: Oh well, that was a partial success.  Failed in Go - looks like the new endorsement policy hadn't been picked up in the first private-data test.  Adding short sleep back in. Not sure what the `--waitForEvent` is doing then.

File: java/src/main/java/org/hyperledger/fabric/client/Proposal.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bestbeforetoday) [reply java/src/main/java/org/hyperledger/fabric/client/Proposal.java:78]: 👍

File: scenario/node/src/fabric.ts (hunk 4 above)
Comment (bestbeforetoday) [reply scenario/node/src/fabric.ts]: My _guess_ is that the `--waitForEvent` is waiting for the config update to be committed, but at that point some of the behind-the-scenes deployment work may not yet be complete

File: java/src/main/java/org/hyperledger/fabric/client/ProposalImpl.java [hunk 6]
```diff
@@ -66,6 +70,7 @@ public Transaction endorse() {
                 .setTransactionId(proposedTransaction.getTransactionId())
                 .setChannelId(channelName)
                 .setProposedTransaction(proposedTransaction.getProposal())
+                .addAllEndorsingOrganizations(Arrays.asList(endorsingOrgs))
```
[the remark this answers is not in the document]
Comment (andrew-coleman) [reply java/src/main/java/org/hyperledger/fabric/client/ProposalImpl.java]: It makes no difference as far as the behaviour of the gateway is concerned.  It treats empty array as unset.

**Review by bestbeforetoday — APPROVED** [review]

**Review by bestbeforetoday — APPROVED** [review]

**Pull request merged**